### PR TITLE
Add subscription limits

### PR DESCRIPTION
**Pull request by cnderrauber**

### CODE REVIEW

File: pkg/rtc/subscriptionmanager.go [hunk 1]
```diff
@@ -50,14 +50,19 @@ type SubscriptionManagerParams struct {
...
 }
 
 // SubscriptionManager manages a participant's subscriptions
 type SubscriptionManager struct {
 	params        SubscriptionManagerParams
 	lock          sync.RWMutex
 	subscriptions map[livekit.TrackID]*trackSubscription
-	subscribedTo  map[livekit.ParticipantID]map[livekit.TrackID]struct{}
+
+	subscribedVideoCount, subscribedAudioCount atomic.Int32
```
Comment (boks1971) [inline pkg/rtc/subscriptionmanager.go:63]: just thinking out aloud, should we have limits per track source?

**Review by boks1971 — APPROVED** [review]

File: pkg/rtc/subscriptionmanager.go (hunk 1 above)
Comment (cnderrauber) [reply pkg/rtc/subscriptionmanager.go:63]: Do you mean camera or screen share? I can't image the scenario to limit at source level now

Comment (boks1971) [reply pkg/rtc/subscriptionmanager.go:63]: yeah, I was thinking source based. We can do it later.

I was thinking about a case of reaching video limit and somebody starts a screen share. I am guessing users would want to receive that. So, having a different limit would help with that. Also, there is not going to be a lot of parallel screen shares. So, that can have a low limit. Of course, we can added it later.

Comment (cnderrauber) [reply pkg/rtc/subscriptionmanager.go:63]: Sounds like a priority, makes sense to have screen share as preferred.

File: pkg/rtc/participant.go [hunk 2]
```diff
@@ -1063,13 +1065,15 @@ func (p *ParticipantImpl) setupUpTrackManager() {
...
-		OnSubscriptionError: p.onSubscriptionError,
+		Participant:            p,
+		Logger:                 p.params.Logger.WithoutSampler(),
+		TrackResolver:          p.params.TrackResolver,
+		Telemetry:              p.params.Telemetry,
+		OnTrackSubscribed:      p.onTrackSubscribed,
+		OnTrackUnsubscribed:    p.onTrackUnsubscribed,
+		OnSubscriptionError:    p.onSubscriptionError,
+		SubscriptionLimitVideo: p.params.SubscriptionLimitVideo,
+		SubscriptionLimitAudio: p.params.SubscriptionLimitAudio,
```
Comment (davidzhao) [inline pkg/rtc/participant.go:1076]: we should make these configurable as a default [room config](https://github.com/livekit/livekit/blob/master/config-sample.yaml#L137) as well. that'll give OSS users a way to customize these settings.

File: pkg/rtc/subscriptionmanager.go [hunk 3]
```diff
@@ -293,13 +298,14 @@ func (m *SubscriptionManager) reconcileSubscription(s *trackSubscription) {
 				if s.durationSinceStart() > subscriptionTimeout {
 					s.maybeRecordError(m.params.Telemetry, m.params.Participant.ID(), err, true)
 				}
-			case ErrTrackNotFound:
+			case ErrTrackNotFound, ErrSubscriptionLimitExceeded:
```
Comment (davidzhao) [inline pkg/rtc/subscriptionmanager.go:301]: we should *not* change intent of the user in this condition. it's important for desired state to match the client.

instead we should avoid attempting subscription of that type if it's over limit.

**Review by davidzhao — CHANGES_REQUESTED** [review]

File: pkg/rtc/subscriptionmanager.go [hunk 4]
```diff
@@ -50,32 +54,34 @@ type SubscriptionManagerParams struct {
...
 }
 
 // SubscriptionManager manages a participant's subscriptions
 type SubscriptionManager struct {
 	params        SubscriptionManagerParams
 	lock          sync.RWMutex
 	subscriptions map[livekit.TrackID]*trackSubscription
-	subscribedTo  map[livekit.ParticipantID]map[livekit.TrackID]struct{}
-	// keeps track of tracks that are already queued for reconcile to avoid duplicating reconcile requests
-	pendingReconcile map[livekit.TrackID]struct{}
```
Comment (cnderrauber) [inline pkg/rtc/subscriptionmanager.go:62]: Remove the map that has never been assigned

File: pkg/rtc/participant.go (hunk 2 above)
Comment (cnderrauber) [reply pkg/rtc/participant.go:1076]: Added to the limits config as the limits are more related to the node resource and don't want config them by the service API.

File: pkg/rtc/participant.go [hunk 5]
```diff
@@ -1763,11 +1767,6 @@ func (p *ParticipantImpl) getPendingTrack(clientId string, kind livekit.TrackTyp
 	if pendingInfo == nil {
 	track_loop:
 		for cid, pti := range p.pendingTracks {
-			if cid == clientId {
```
Comment (cnderrauber) [inline pkg/rtc/participant.go:1766]: remove these as the `clientID` key is already checked above

File: config-sample.yaml [hunk 6]
```diff
@@ -261,3 +261,9 @@ keys:
 #   num_tracks: -1
 #   # defaults to 1 GB/s, or just under 10 Gbps
 #   bytes_per_sec: 1_000_000_000
+#   # how many maximum tracks (audio / video) that a single participant can subscribe at same time.
+#   # if the limits is exceeded, subscriptions will be pending until any subscribed track has been unsubscribed.
+#   # value less or equal than 0 means no limits.
+#   subscription_limits_video: 0
```
Comment (davidzhao) [inline config-sample.yaml:267]: I think one of these should be audio

`subscription_limit_video`
`subscription_limit_audio`

File: pkg/config/config.go [hunk 7]
```diff
@@ -246,8 +246,10 @@ type RegionConfig struct {
 }
 
 type LimitConfig struct {
-	NumTracks   int32   `yaml:"num_tracks,omitempty"`
-	BytesPerSec float32 `yaml:"bytes_per_sec,omitempty"`
+	NumTracks               int32   `yaml:"num_tracks,omitempty"`
+	BytesPerSec             float32 `yaml:"bytes_per_sec,omitempty"`
+	SubscriptionLimitsVideo int32   `yaml:"subscription_limits_video,omitempty"`
```
Comment (davidzhao) [inline pkg/config/config.go:251]: nit: `SubscriptionLimitVideo`

File: pkg/rtc/subscriptionmanager.go [hunk 8]
```diff
@@ -392,13 +391,32 @@ func (m *SubscriptionManager) reconcileWorker() {
 	}
 }
 
+func (m *SubscriptionManager) hasCapcityForSubscribption(kind livekit.TrackType) bool {
```
Comment (davidzhao) [inline pkg/rtc/subscriptionmanager.go:394]: typo here
```suggestion
func (m *SubscriptionManager) hasCapcityForSubscription(kind livekit.TrackType) bool {
```

**Review by davidzhao — APPROVED** [review]
awesome work!

**Pull request merged**